Improve request building with an alloy-style filler system

nategraf opened the pull request:
Inspired by `alloy`'s `TxFiller`, this draft PR is the beginning of an attempt to improve request building using a similar approach.

willpote commented on crates/boundless-market/src/request_builder/finalizer.rs:
```diff
@@ -0,0 +1,121 @@
+        // As an extra consistency check. verify the journal satisfies the required predicate.
+        if let Some(ref journal) = self.journal {
+            if !requirements.predicate.eval(journal) {
+                bail!("journal in request builder does not match requirements predicate; check request parameters");
```
Lets print the type, provided, and the expected values. We had an issue where Hans was trying to use this and it was hard to debug

willpote reviewed (approved):
Very cool!

nategraf commented on crates/boundless-market/src/request_builder/mod.rs:
````diff
@@ -0,0 +1,719 @@
+    /// ```
+    ///
+    /// See also [Self::with_env] and [GuestEnvBuilder][crate::input::GuestEnvBuilder]
+    pub fn with_stdin(self, value: impl Into<Vec<u8>>) -> Self {
````
Mainly that we have the `Input` struct that, which this is only a part of. We could rename that struct though if you have an ideas there. Or just say that this potential confusion is not an issue, and that confusion from calling this `stdin` is worse.

willpote replied:
Hmm yeah I now notice it looks like we've already renamed the Solidity `Input` struct in this file to "RequestInput" I suppose to try and reduce potential for confusion?

One idea would be to keep the Solidity/Rust versions both as `Input` to minimize confusion, and rename `with_stdin` to `with_input_bytes`

willpote replied:
Also side question, what happens if a user sets both `with_stdin` and `with_request_input` at the moment? Which ends up taking precedence?

nategraf replied:
Based on our discussion earlier, I think we are going to stick with this API for now. I am working right now renaming `Input` to `RequestInput` in the Rust codebase.

And to your question,`with_request_input` will superseded the over input methods. This is handled in the storage layer. It could be done better, to make this harder to mess up, but I think this is probably as good as its going to be for now. I've added a comment about this interaction in a9722fd0

nategraf commented on crates/boundless-market/src/client.rs:
```diff
@@ -334,32 +513,60 @@ where
 
-    /// Submit a proof request.
+    /// Initial parameters that will be used to build a [ProofRequest] using the [RequestBuilder].
+    pub fn request_params<Params>(&self) -> Params
```
Ya, its not perfect. I don't love `create_request` because it is unclear that its not actually building the request right away, and is unclear how it is distinguished from `client.build_request` (which uses the `RequestBuilder`).

I thought about `start_request` or something like that. Not sure.

nategraf replied:
In 851ca81d828f13fadd097fe3e5a616e0b8c9fa2f I renamed this to `new_request`. What do you think of that? This is the best I have thought of. I had some concern with `start_request` that folks might think its starting some kind of subprocess or something, and that is explicitly not the case.

nategraf commented on crates/boundless-market/src/request_builder/mod.rs:
```diff
@@ -0,0 +1,719 @@
+    type Error;
+
+    async fn process_with(self, layer: &L) -> Result<Self::Output, Self::Error>;
+}
```
We'll see if all this Rust type system work pays off. The point is to allow external devs to create a custom `RequestBuilder` and use it directly in the `Client` without having to rewrite the common logic (e.g. for preflight or storage).

nategraf commented:
@willpote I believe I've addressed all your comments, and this is now passing all the tests I've run locally, so I think it's ready to go. If my responses look good, feel free to hit merge on this.

nategraf commented on crates/boundless-market/src/request_builder/mod.rs:
```diff
@@ -0,0 +1,719 @@
+    }
+
+    pub fn require_env(&self) -> Result<&GuestEnv, MissingFieldError> {
+        self.env.as_ref().ok_or(MissingFieldError::new("env"))
```
Oop, I missed addressing this one until doing a last look. I've added some hints and context to errors in 84291fe6a1e29481d2ff3a575717c5e2fd3d78b6
